File: ProjectEuler/551-600/S587.cs
using System;

namespace ProjectEuler
{
    public class S587 : ISolution
    {
        public string GetAnswer()
        {
            int n = 16;
            while (true)
            {
                double y = (n - Math.Sqrt(2 * n) + 1) / (n * n + 1);
                double x = y * n;
                double sTriangle = x * y / 2;
                double sTrapezoid = (y + 1) * (1 - x) / 2;
                double theta = Math.Atan((1 - x) / (1 - y));
                double sConcaveTriangle = sTriangle + sTrapezoid - theta / 2;
                double sL = 1 - Math.PI / 4;
                if (sL / sConcaveTriangle > 1000)
                {
                    break;
                }

                n++;
            }

            return n.ToString();
        }
    }
}

File: src/551-600/S587.cc
#include "S587.h"
#include <cmath>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

namespace ProjectEuler
{
std::string S587::GetAnswer()
{
	int n = 16;
	while (true)
	{
		double y = (n - std::sqrt(2 * n) + 1) / (n * n + 1);
		double x = y * n;
		double sTriangle = x * y / 2;
		double sTrapezoid = (y + 1) * (1 - x) / 2;
		double theta = std::atan((1 - x) / (1 - y));
		double sConcaveTriangle = sTriangle + sTrapezoid - theta / 2;
		double sL = 1 - M_PI / 4;
		if (sConcaveTriangle / sL < 0.001)
		{
			break;
		}

		n++;
	}

	return std::to_string(n);
}
} // namespace ProjectEuler
